src/message/pack_signed.rs

```diff
@@ -26,52 +36,226 @@ impl Message {
     /// - `DIDUrlNotResolved` DID doesn't contain mentioned DID Urls (for ex., key id)
     /// - `SecretNotFound` Sender secret is not found.
+    /// - `Unsupported` Used crypto or method is unsupported.
     /// - `InvalidState` Indicates library error.
     /// - `IOError` IO error during DID or secrets resolving
     /// TODO: verify and update errors list
     pub async fn pack_signed<'dr, 'sr>(
         &self,
-        _sign_by: &str,
-        _did_resolver: &'dr (dyn DIDResolver + 'dr),
-        _secrets_resolver: &'sr (dyn SecretsResolver + 'sr),
+        sign_by: &str,
+        did_resolver: &'dr (dyn DIDResolver + 'dr),
+        secrets_resolver: &'sr (dyn SecretsResolver + 'sr),
     ) -> Result<(String, PackSignedMetadata)> {
-        todo!()
+        let (did, key_id) = did_or_url(sign_by);
+
+        let did_doc = did_resolver
+            .resolve(did)
+            .await
+            .context("Unable resolve signer did")?
+            .ok_or_else(|| err_msg(ErrorKind::DIDNotResolved, "Signer did not found"))?;
+
+        let authentications: Vec<_> = if let Some(key_id) = key_id {
+            did_doc
+                .authentications
+                .iter()
+                .find(|a| *a == key_id)
+                .ok_or_else(|| {
+                    err_msg(
+                        ErrorKind::DIDUrlNotFound,
+                        "Signer key id not found in did doc",
+                    )
+                })?;
+
+            vec![key_id]
+        } else {
+            did_doc.authentications.iter().map(|s| s.as_str()).collect()
+        };
+
+        let key_id = *secrets_resolver
+            .find_secrets(&authentications)
+            .await
+            .context("Unable find secrets")?
+            .get(0)
+            .ok_or_else(|| err_msg(ErrorKind::SecretNotFound, "No signer secrets found"))?;
+
+        let secret = secrets_resolver
+            .get_secret(key_id)
+            .await
+            .context("Unable get secret")?
+            .ok_or_else(|| err_msg(ErrorKind::SecretNotFound, "Signer secret not found"))?;
+
+        let sign_key = secret
+            .to_sign_key_pair()
+            .context("Unable instantiate sign key")?;
+
+        let payload = serde_json::to_string(self)
+            .kind(ErrorKind::InvalidState, "Unable serialize message")?;
+
+        let msg = match sign_key {
+            SignKeyPair::Ed25519KeyPair(ref key) => {
+                jws::sign(payload.as_bytes(), (key_id, key), Algorithm::EdDSA)
+            }
+            SignKeyPair::P256KeyPair(ref key) => {
+                jws::sign(payload.as_bytes(), (key_id, key), Algorithm::Es256)
+            }
+            SignKeyPair::K256KeyPair(ref key) => {
+                jws::sign(payload.as_bytes(), (key_id, key), Algorithm::Es256K)
+            }
+        }
+        .context("Unable produce signatire")?;
+
+        let metadata = PackSignedMetadata {
+            sign_by_kid: key_id.to_owned(),
```

Comment:
In Kotlin \ Python a signed message is returned in Metadata as well

Reply:
In Kotlin \ Python this parameter named sign_from

Reply:
Let's keep it as is and discuss with Alexa and team next week. I like my naming better